pkg/handlers/deploy.go

```diff
@@ -140,5 +166,144 @@ func (h *Handler) DeployAppVersion(w http.ResponseWriter, r *http.Request) {
 
 	deployAppVersionResponse.Success = true
 
-	JSON(w, 204, deployAppVersionResponse)
+	JSON(w, http.StatusOK, deployAppVersionResponse)
+}
+
+func isKotsUpgradeNeeded(app *apptypes.App, sequence int64) (bool, string, error) {
+	archivePath, err := ioutil.TempDir("", "kotsadm")
+	if err != nil {
+		return false, "", errors.Wrap(err, "failed to create temp dir")
+	}
+	defer os.RemoveAll(archivePath)
+
+	err = store.GetStore().GetAppVersionArchive(app.ID, int64(sequence), archivePath)
+	if err != nil {
+		return false, "", errors.Wrapf(err, "failed to get archive for sequence %d", sequence)
+	}
+
+	kotsKinds, err := kotsutil.LoadKotsKindsFromPath(archivePath)
+	if err != nil {
+		return false, "", errors.Wrap(err, "failed to load kots kinds")
+	}
+
+	latestVersion, _ := findLatestKotsVersion(app.ID, kotsKinds.License)
+
+	targetVersion, err := getTargetKotsVersion(kotsKinds, latestVersion)
+	if err != nil {
+		if err, ok := err.(AdminConsoleUpgradeError); ok {
+			if err.IsCritical {
+				return false, "", errors.Wrap(err, "critical error while checking target kots version")
+			}
+			logger.Infof("skipping kots upgrade because %v", err)
+			return false, "", nil
+		}
+		return false, "", errors.Wrap(err, "failed to find target version")
+	}
+
+	return true, targetVersion, nil
+}
+
+func findLatestKotsVersion(appID string, license *kotsv1beta1.License) (string, error) {
+	url := fmt.Sprintf("%s/admin-console/version/latest", license.Spec.Endpoint)
+
+	req, err := http.NewRequest("GET", url, nil)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to create new request")
+	}
+
+	reportingInfo := reporting.GetReportingInfo(appID)
+	reporting.InjectReportingInfoHeaders(req, reportingInfo)
+
+	req.Header.Add("User-Agent", fmt.Sprintf("KOTS/%s", buildversion.Version()))
+	req.Header.Set("Authorization", fmt.Sprintf("Basic %s", base64.StdEncoding.EncodeToString([]byte(fmt.Sprintf("%s:%s", license.Spec.LicenseID, license.Spec.LicenseID)))))
+
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to execute get request")
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to read response body")
+	}
+
+	if resp.StatusCode >= 400 {
+		if len(body) > 0 {
+			return "", util.ActionableError{Message: string(body)}
+		}
+		return "", errors.Errorf("unexpected result from get request: %d", resp.StatusCode)
+	}
+
+	var versionInfo struct {
+		Tag string `json:"tag"`
+	}
+	if err := json.Unmarshal(body, &versionInfo); err != nil {
+		return "", errors.Wrapf(err, "failed to unmarshal response: %s", body)
+	}
+
+	return versionInfo.Tag, nil
+}
+
+type AdminConsoleUpgradeError struct {
+	IsCritical bool
+	Message    string
+}
+
+func (e AdminConsoleUpgradeError) Error() string {
+	return e.Message
+}
+
+func getTargetKotsVersion(kotsKinds *kotsutil.KotsKinds, latestVersion string) (string, error) {
```

Comment:
why not just use the existing `IsKotsVersionCompatibleWithApp` function?

Reply:
I don't really see how it's the same :)